### PR TITLE
Conn ip information, keeping it together

### PR DESCRIPTION
**Pull request by icing**
new struct ip_quadruple for holding local/remote addr+port

- used in data->info and conn and cf-socket.c
- copy back and forth complete struct
- add 'secondary' to conn
- use secondary in reporting success for ftp 2nd connection
- refs #13084

### CODE REVIEW

**Comment by bagder** [comment]
I propose this additional patch to use plain struct assignments instead of mempy on a few places:
~~~diff
From 4bd96f431b2bdb2fdf0dc823e9561fdccab52d44 Mon Sep 17 00:00:00 2001
From: Daniel Stenberg <daniel@haxx.se>
Date: Fri, 8 Mar 2024 12:57:14 +0100
Subject: [PATCH] reduce memcpy

---
 lib/cf-socket.c | 6 +++---
 lib/connect.c   | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/lib/cf-socket.c b/lib/cf-socket.c
index 4acb09677..a32d4b4b1 100644
--- a/lib/cf-socket.c
+++ b/lib/cf-socket.c
@@ -1439,13 +1439,13 @@ static void conn_set_ip(struct Curl_cfilter *cf, struct Curl_easy *data)
 {
   struct cf_socket_ctx *ctx = cf->ctx;
 
   (void)data;
   if(cf->sockindex == SECONDARYSOCKET)
-    memcpy(&cf->conn->secondary, &ctx->ip, sizeof(cf->conn->secondary));
+    cf->conn->secondary = ctx->ip;
   else
-    memcpy(&cf->conn->primary, &ctx->ip, sizeof(cf->conn->primary));
+    cf->conn->primary = ctx->ip;
 }
 
 static void cf_socket_active(struct Curl_cfilter *cf, struct Curl_easy *data)
 {
   struct cf_socket_ctx *ctx = cf->ctx;
@@ -1973,10 +1973,10 @@ CURLcode Curl_cf_socket_peek(struct Curl_cfilter *cf,
     if(psock)
       *psock = ctx->sock;
     if(paddr)
       *paddr = &ctx->addr;
     if(pip)
-      memcpy(pip, &ctx->ip, sizeof(*pip));
+      *pip = ctx->ip;
     return CURLE_OK;
   }
   return CURLE_FAILED_INIT;
 }
diff --git a/lib/connect.c b/lib/connect.c
index 1a0f64ed2..e457006c0 100644
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -146,11 +146,11 @@ timediff_t Curl_timeleft(struct Curl_easy *data,
    the transfer handle is no longer associated with the connection. */
 void Curl_persistconninfo(struct Curl_easy *data, struct connectdata *conn,
                           struct ip_quadruple *ip)
 {
   if(ip)
-    memcpy(&data->info.primary, ip, sizeof(data->info.primary));
+    data->info.primary = *ip;
   else {
     memset(&data->info.primary, 0, sizeof(data->info.primary));
     data->info.primary.remote_port = -1;
     data->info.primary.local_port = -1;
   }
-- 
2.43.0

**Comment by icing** [comment]
Thanks! Added.

File: lib/cf-socket.c [hunk 1]
```diff
@@ -1458,15 +1458,17 @@ static void cf_socket_active(struct Curl_cfilter *cf, struct Curl_easy *data)
...
-    Curl_persistconninfo(data, cf->conn, ctx->l_ip, ctx->l_port);
+    Curl_persistconninfo(data, cf->conn, &ctx->ip);
     /* buffering is currently disabled by default because we have stalls
      * in parallel transfers where not all buffered data is consumed and no
      * socket events happen.
      */
     ctx->buffer_recv = FALSE;
   }
+  else {
+  }
```
Comment (bagder) [inline lib/cf-socket.c:1471]: maybe drop this?

Comment (icing) [reply lib/cf-socket.c]: Reshuffled this in latest push.

**Pull request closed without merging**